### PR TITLE
Fail COM CoClass creation during construction instead of during jitting

### LINKED ISSUE #54379: Test failure: Interop\\COM\\NETClients\\ComDisabled\\NETClientComDisabled\\NETClientComDisabled.cmd
https://dev.azure.com/dnceng/public/_build/results?buildId=1191223&view=ms.vss-test-web.build-test-results-tab&runId=35762252&resultId=101048&paneView=debug

R2R-CG2 windows x64 Checked no_tiered_compilation

(and other legs)

```
    Interop\COM\NETClients\ComDisabled\NETClientComDisabled\NETClientComDisabled.cmd [FAIL]
      Could Not Find C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\COMNativeServer.dll.rsp
      Could Not Find C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\NETClientComDisabled.dll.rsp
      Could Not Find C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\TestLibrary.dll.rsp
      Unhandled exception. System.NotSupportedException: Built-in COM has been disabled via a feature switch. See https://aka.ms/dotnet-illink/com for more information.
         at NetClient.Program.ActivateServer() in NETClientComDisabled.dll:token 0x600007f+0x0
         at NetClient.Program.Main(String[] doNotUse) in NETClientComDisabled.dll:token 0x600007e+0xb
      
      Return code:      1
      Raw output file:      C:\h\w\A9AB0918\w\A6780901\uploads\Reports\Interop.COM\NETClients\ComDisabled\NETClientComDisabled\NETClientComDisabled.output.txt
      Raw output:
      BEGIN EXECUTION
      COMNativeServer.dll
      NETClientComDisabled.dll
      TestLibrary.dll
              3 file(s) copied.
      Response file: C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\COMNativeServer.dll.rsp
      C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\COMNativeServer.dll
      -o:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\COMNativeServer.dll
      --targetarch:x64
      --verify-type-and-field-layout
      -r:C:\h\w\A9AB0918\p\System.*.dll
      -r:C:\h\w\A9AB0918\p\Microsoft.*.dll
      -r:C:\h\w\A9AB0918\p\mscorlib.dll
      -r:C:\h\w\A9AB0918\p\netstandard.dll
      -O
      " "dotnet" "C:\h\w\A9AB0918\p\crossgen2\crossgen2.dll" @"C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\COMNativeServer.dll.rsp"   -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll"
      No input files are loadable
      Response file: C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\NETClientComDisabled.dll.rsp
      C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\NETClientComDisabled.dll
      -o:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\NETClientComDisabled.dll
      --targetarch:x64
      --verify-type-and-field-layout
      -r:C:\h\w\A9AB0918\p\System.*.dll
      -r:C:\h\w\A9AB0918\p\Microsoft.*.dll
      -r:C:\h\w\A9AB0918\p\mscorlib.dll
      -r:C:\h\w\A9AB0918\p\netstandard.dll
      -O
      " "dotnet" "C:\h\w\A9AB0918\p\crossgen2\crossgen2.dll" @"C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\NETClientComDisabled.dll.rsp"   -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll"
      Emitting R2R PE file: C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\NETClientComDisabled.dll
      Response file: C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\TestLibrary.dll.rsp
      C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\TestLibrary.dll
      -o:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\TestLibrary.dll
      --targetarch:x64
      --verify-type-and-field-layout
      -r:C:\h\w\A9AB0918\p\System.*.dll
      -r:C:\h\w\A9AB0918\p\Microsoft.*.dll
      -r:C:\h\w\A9AB0918\p\mscorlib.dll
      -r:C:\h\w\A9AB0918\p\netstandard.dll
      -O
      " "dotnet" "C:\h\w\A9AB0918\p\crossgen2\crossgen2.dll" @"C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\TestLibrary.dll.rsp"   -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll -r:C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\IL-CG2\*.dll"
      Emitting R2R PE file: C:\h\w\A9AB0918\w\A6780901\e\Interop\COM\NETClients\ComDisabled\NETClientComDisabled\\TestLibrary.dll
       "C:\h\w\A9AB0918\p\corerun.exe" -p "System.Runtime.InteropServices.BuiltInComInterop.IsSupported=false" NETClientComDisabled.dll 
      Expected: 100
      Actual: -532462766
      END EXECUTION - FAILED
      FAILED
```

Related? https://github.com/dotnet/runtime/pull/54298

@jkoritzinsky @LakshanF

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fail COM CoClass creation during construction instead of during jitting (which wouldn't fail when the assembly had been run through crossgen2). Also, make the behavior consistent across platforms.

Fixes https://github.com/dotnet/runtime/issues/54379

### CODE REVIEW

**Comment by jkotas** [comment]
Re-enable the test disabled in https://github.com/dotnet/runtime/pull/54435/

File: src/coreclr/vm/gchelpers.cpp [hunk 1]
```diff
@@ -930,15 +930,25 @@ OBJECTREF AllocateObject(MethodTable *pMT
...
             COMPlusThrow(kInvalidComObjectException, IDS_EE_NO_BACKING_CLASS_FACTORY);
 
         oref = OBJECTREF_TO_UNCHECKED_OBJECTREF(AllocateComObject_ForManaged(pMT));
     }
-    else
 #endif // FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+#else  // FEATURE_COMINTEROP
+    if (pMT->IsComObjectType())
+    {
+        COMPlusThrow(kNotSupportedException, IDS_EE_ERROR_COM);
```
Comment (elinor-fung) [inline src/coreclr/vm/gchelpers.cpp:948]: Should this stay `PlatformNotSupportedException`? I think that would be more consistent with the current split of `NotSupportedException` for cases where COM support is disabled and  `PlatformNotSupportedException` for cases where COM support doesn't exist.

**Review by elinor-fung — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Failure is a known infra issue. Merging this in.

**Pull request merged**